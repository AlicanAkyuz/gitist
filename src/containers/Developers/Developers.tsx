import React, { FunctionComponent } from 'react';
import { Query } from 'react-apollo';
import { GET_DEVELOPERS } from '../../queries';

import Developer from '../../components/Developer';
import Loading from '../../components/Loading';
import Error from '../../components/Error';

interface Data {
  location: {
    stats: {
      rank: number;
    };
    name: string;
    developers: [
      {
        totalStarred: number;
        name: string;
        followers: number;
        avatarUrl: string;
      },
    ];
  };
}

interface Variables {
  slug: string | null;
  limit: number;
  offset: number;
  orderBy: object;
}

const Developers: FunctionComponent = () => {
  let city = localStorage.getItem('city');
  return (
    <Query<Data, Variables>
      query={GET_DEVELOPERS}
      variables={{
        slug: city,
        orderBy: { direction: 'DESC', field: 'TOTAL_STARRED' },
        limit: 5,
        offset: 0,
      }}
    >
      {({ data, loading, error, fetchMore }) => {
        if (loading || !data) return <Loading />;
        if (error) return <Error />;
        const { location } = data;
        return <Developer location={location} />;
      }}
    </Query>
  );
};

export default Developers;
