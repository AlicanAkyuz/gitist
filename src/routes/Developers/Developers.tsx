import React, { FunctionComponent } from 'react';
import { Query } from 'react-apollo';
import { GET_DEVELOPERS } from '../../queries';

import DeveloperList from '../../components/DeveloperList';
import Loading from '../../components/Loading';
import Error from '../../components/Error';

interface Data {
  location: {
    stats: {
      rank: number;
    };
    name: string;
    developers: [
      {
        totalStarred: number;
        name: string;
        followers: number;
        avatarUrl: string;
      },
    ];
  };
}

interface Variables {
  slug: string | null;
  limit: number;
  offset: number | [];
  orderBy: object;
}

const Developers: FunctionComponent = () => {
  let city = localStorage.getItem('city');
  return (
    <Query<Data, Variables>
      query={GET_DEVELOPERS}
      variables={{
        slug: city,
        orderBy: { direction: 'DESC', field: 'TOTAL_STARRED' },
        limit: 5,
        offset: 0,
      }}
    >
      {({ data, loading, error, fetchMore }) => {
        if (error) return <Error />;
        if (loading || !data) return <Loading />;
        let offset = 0;
        const { location } = data;
        return (
          <DeveloperList
            location={location}
            onLoadMore={async () => {
              try {
                offset = offset + 5;
                await fetchMore({
                  variables: { offset },
                  updateQuery: (prev, { fetchMoreResult }) => {
                    if (!fetchMoreResult) return prev;
                    let newData = JSON.parse(JSON.stringify(prev));
                    newData.location.developers = Object.assign(
                      [],
                      [...prev.location.developers, ...fetchMoreResult.location.developers],
                    );
                    return newData;
                  },
                });
              } catch (err) {
                console.log('err');
                // When the variables of <Query> changes while a fetchMore is still in progress, we get this error: ObservableQuery with this id does not exist: id on unmounted component, which is an open issue: https://github.com/apollographql/apollo-client/issues/4114
              }
            }}
          />
        );
      }}
    </Query>
  );
};

export default Developers;
