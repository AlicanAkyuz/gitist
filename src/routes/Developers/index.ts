import Developers from './Developers';

export default Developers;
