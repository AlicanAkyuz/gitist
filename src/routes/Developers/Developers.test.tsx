import React from 'react';
import { shallow } from 'enzyme';

import Developers from './Developers';

it('Developers corrently renders a navbar and Routes component', () => {
  const wrapper = shallow(<Developers />);
  expect(wrapper).toMatchSnapshot();
});
