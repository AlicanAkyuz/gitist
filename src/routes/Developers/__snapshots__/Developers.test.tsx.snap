// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Developers corrently renders a navbar and Routes component 1`] = `
<Query
  query={
    Object {
      "definitions": Array [
        Object {
          "directives": Array [],
          "kind": "OperationDefinition",
          "name": Object {
            "kind": "Name",
            "value": "location",
          },
          "operation": "query",
          "selectionSet": Object {
            "kind": "SelectionSet",
            "selections": Array [
              Object {
                "alias": undefined,
                "arguments": Array [
                  Object {
                    "kind": "Argument",
                    "name": Object {
                      "kind": "Name",
                      "value": "slug",
                    },
                    "value": Object {
                      "kind": "Variable",
                      "name": Object {
                        "kind": "Name",
                        "value": "slug",
                      },
                    },
                  },
                ],
                "directives": Array [],
                "kind": "Field",
                "name": Object {
                  "kind": "Name",
                  "value": "location",
                },
                "selectionSet": Object {
                  "kind": "SelectionSet",
                  "selections": Array [
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "name",
                      },
                      "selectionSet": undefined,
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "stats",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "rank",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                    Object {
                      "alias": undefined,
                      "arguments": Array [
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "orderBy",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "orderBy",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "limit",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "limit",
                            },
                          },
                        },
                        Object {
                          "kind": "Argument",
                          "name": Object {
                            "kind": "Name",
                            "value": "offset",
                          },
                          "value": Object {
                            "kind": "Variable",
                            "name": Object {
                              "kind": "Name",
                              "value": "offset",
                            },
                          },
                        },
                      ],
                      "directives": Array [],
                      "kind": "Field",
                      "name": Object {
                        "kind": "Name",
                        "value": "developers",
                      },
                      "selectionSet": Object {
                        "kind": "SelectionSet",
                        "selections": Array [
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "name",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "avatarUrl",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "totalStarred",
                            },
                            "selectionSet": undefined,
                          },
                          Object {
                            "alias": undefined,
                            "arguments": Array [],
                            "directives": Array [],
                            "kind": "Field",
                            "name": Object {
                              "kind": "Name",
                              "value": "followers",
                            },
                            "selectionSet": undefined,
                          },
                        ],
                      },
                    },
                  ],
                },
              },
            ],
          },
          "variableDefinitions": Array [
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "String",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "slug",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NonNullType",
                "type": Object {
                  "kind": "NamedType",
                  "name": Object {
                    "kind": "Name",
                    "value": "DeveloperOrder",
                  },
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "orderBy",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "Int",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "limit",
                },
              },
            },
            Object {
              "defaultValue": undefined,
              "directives": Array [],
              "kind": "VariableDefinition",
              "type": Object {
                "kind": "NamedType",
                "name": Object {
                  "kind": "Name",
                  "value": "Int",
                },
              },
              "variable": Object {
                "kind": "Variable",
                "name": Object {
                  "kind": "Name",
                  "value": "offset",
                },
              },
            },
          ],
        },
      ],
      "kind": "Document",
      "loc": Object {
        "end": 322,
        "start": 0,
      },
    }
  }
  variables={
    Object {
      "limit": 5,
      "offset": 0,
      "orderBy": Object {
        "direction": "DESC",
        "field": "TOTAL_STARRED",
      },
      "slug": null,
    }
  }
>
  <Component />
</Query>
`;
